### ISSUE TITLE
Type mismatch using dot-sourcing

### ISSUE THREAD

**Issue by ITCTOM** [PowerShell/PowerShell#20273]
### Prerequisites

- [x] Write a descriptive title.
- [x] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [x] Search the existing issues.
- [x] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [x] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

My test-scriptfile:
```powershell
@'
bla
blabla
---end---
bla
blabla
'@ > c:\temp\testfile.txt

function searching_with_select-string {
    $content = Get-Content c:\temp\testfile.txt
    [string]$start= $content | select-string 'Start' | Select-Object -last 1
    [string]$end = $content | Select-String 'End' | Select-Object -Last 1
    write-host "Start: $start"
    write-host "End: $end"
    write-host "Type Start: $($start.gettype())"
    write-host "Type End: $($end.gettype())"
}

Write-Host "First functionscall without dot-sourcing searching_with_select-string"
searching_with_select-string
Write-Host "Second functioncall with dot-sourcing"
. searching_with_select-string
``` 

Call the function once without dot-sourcing and with dot-sourcing
Regardless of whether the function is called using dot-sourcing or not, I would expect the same output. But I get different output

### Expected behavior

```console
I would expect that $start is in both cases $null because the string "Start" isn´t found in the testtextfile.txt So I expect that I get an error message when I call $start.gettype() because $start is $null and $null has no datetype.
But I get a diffrent Output in both function calls.
```

### Actual behavior

```console
PS C:\Users\ad006pa> searching_with_select-string  
Start:        
End: ---end---
InvalidOperation: C:\temp\test1.ps1:15:31
Line |
  15 |      write-host "Type Start: $($start.gettype())"    
     |                                ~~~~~~~~~~~~~~~~      
     | You cannot call a method on a null-valued expression.
Type Start:
Type End: string
PS C:\Users\ad006pa> . searching_with_select-string
Start:
End: ---end---       
Type Start: string   
Type End: string
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.6
PSEdition                      Core
GitCommitId                    7.3.6
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by mklement0** [comment]
I think you're seeing a variation of the problem I just stumbled across in the context of strongly typed _array_ variables:

 * #20275

Minimal repro:

```powershell
# Run interactively, in the global scope; & {} is a dummy command that produces no output, i.e.
# technically the [System.Management.Automation.Internal.AutomationNull]::Value singleton.
$null -eq ([string] $start = & {}); $start -is [string]
'---'
# Now run in a child scope
& { $null -eq ([string] $start = & {}); $start -is [string] }
```

You'll get `$false, $true` in the first case and `$true, $false` in the second.
That is, in the _global_ scope the _empty string_ was assigned, and in the non-global one `$null`

However, whereas in #20275 it seems that consistently assigning `$null` is the appropriate solution, 
here you could argue the opposite: 

 * Given that `[string]` even converts `$null` to `''` (the empty string) - e.g., `([string] $null) -is [string]` is _always_ `$true` - you could argue that assigning the empty string is the right behavior (which currently only happens in the _global_ scope).

**Comment by jhoneill** [comment]
@ITCTOM 
I'll offer this version of your script as a way of shedding a little more light on what is happening, with any explanation for **why** it should be that way because the inconsistency is confusing

```
function searching_with_select-string {
    $content = Get-Content c:\temp\testfile.txt
     [string]$start= $content | select-string 'Start' | Select-Object -last 1
     [string]$end = $content | Select-String 'End' | Select-Object -Last 1
     get-variable start | fl *
    $null -eq (get-variable start).value
     $start = 1
     get-variable start | fl *
     $start.gettype()
 }
```

I was going to comment on "strong typing" vs the type conversion attribute, because `[string]` doesn't say "_start_ is a string" it says "convert anything placed in _Start_ into a string", and the distinction doesn't matter to many people much of the time..
 But the conversion attribute behaviour changes in the different scopes.    When it isn't dot sourced there is no type conversion attribute **visible.** Yet assigning 1 to _start_ does convert it from an int to a string so functionality is there (same if you create have [int]$i = 1 and then try $i = "one"  - the attribute might be invisible but there's still a 
'Cannot convert value "one" to type "System.Int32". Error: "The input string 'one' was not in a correct format." '

This behaviour does seem to be in all versions...

**Comment by SeeminglyScience** [comment]
Yeah, more specifically it's the difference between an access of an optimized local vs any other variable access. The former generates a strongly typed property in a mutable tuple and adds a dynamic call to `PSConvertBinder` for every assignment. The latter just calls `VariableOps.SetVariableValue`, which appears to handle `AutomationNull` different.

I'm not entirely sure which is the actually desired behavior though. I guess for string specifically it would make sense for both to set the value as an empty string here, as it's somewhat faked by the engine to appear like a value type. Opening up to the engine WG

**Comment by ITCTOM** [comment]
In my example, neither global variables nor the debugger are used. I suspect that the case in my example is slightly different than the examples discussed here in the comments.

**Comment by mklement0** [comment]
You're right, @ITCTOM:

It isn't about the global scope per se, but - to quote @SeeminglyScience (emphasis added)  - the "difference between an access of an _optimized local_ vs _any other variable access_" and it seems that the latter occurs in (at least) the following scenarios:

* variable access in the _global scope_ (my minimal repro above and the one in the linked issue)
* variable access _in dot-sourced code_ (your repro).

As such, I _assumed_ that #20275 has the same root cause, but my understanding is superficial here, which is why I initially asked for confirmation.

The debugger issue is only mentioned in #20275, but it equally affects your code. I mentioned it in preparation for potentially closing #20275 as a duplicate of _this_ issue.

However, I've now:

* closed #20275 as a duplicate of this one.

* created a separate issue for the debugger problem, as it may affect other behaviors too: #20298

**Comment by ITCTOM** [comment]
i don´t use any dugger instance. This code (see my initial post) works in any Powershell (ecpecially without starting a debugger session) with the described type mismatch.

**Comment by mklement0** [comment]
@ITCTOM, I know, and I didn't mean to imply that you used the debugger (just that if you _did_ use it, the behavior of your code too would change in subsequent calls) - now that debugger issue has been spun out into #20298, we don't need to discuss it here anymore.

**Comment by SeeminglyScience** [comment]
Yes it's about whether the compiler is set to optimize local variable access. This is disabled when dot sourcing, when accessing all scoped variables, when any `*-Debugger` or `*-Variable` command is statically present in the body, and I believe when breakpoints are set. I likely missed a scenario or two, but that's the root cause for all of the differences.

**Comment by SeeminglyScience** [comment]
The Engine WG reviewed this and agree it's a bug and both optimized and non-optimized should handle the default value as an empty string.

**Comment by mklement0** [comment]
Glad to hear it, @SeeminglyScience.
So that this doesn't fall through the cracks, it should be labelled `Issue-Bug`.

Speaking of: 

Let me stress again that it's important to _fundamentally_ harmonize how values are assigned (converted) in the optimized vs. non-optimized scenarios, which should also include preventing AutomationNull from being assigned to an `[array]` type-constrained variable, as originally reported in #20275 (the latter is about _strongly_ typed arrays, but I've come to realize it equally applies to `[array]` aka `[object[]]` arrays too), and should assign `$null` instead.

**Comment by mklement0** [comment]
@iSazonov, I just noticed that this issue is still not labeled `Issue-Bug`, despite the [WG's finding and decision](https://github.com/PowerShell/PowerShell/issues/20273#issuecomment-1743624803) - can you please add it?

Also, it occurred to me that the following long-standing bug report may be related:

* #4312

**Comment by mklement0** [comment]
@SeeminglyScience, given that you agreed that the issue at hand is a bug, can you please label it as such (`Issue-Bug`) and also add the `WG-Reviewed` one?

**Comment by ITCTOM** [comment]
afer nearly one year, thanks a lot labeling this issue as a bug. Looking forward that somesone solves this issue.

**Comment by ITCTOM** [comment]
What´s going on with this issue?
